pkg/asset/installconfig/gcp/usertags.go

```diff
@@ -0,0 +1,298 @@
+package gcp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	rscmgr "cloud.google.com/go/resourcemanager/apiv3"
+	rscmgrpb "cloud.google.com/go/resourcemanager/apiv3/resourcemanagerpb"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/gcp"
+)
+
+const (
+	// maxUserTagLimit is the maximum userTags that can be configured as defined in openshift/api.
+	// https://github.com/openshift/api/commit/ae73a19d05c35068af16c9aeff375d0b7c936a8a#diff-07b264a49084976b670fb699badaca1795027d6ea732a99226a5388104f6174fR604-R613
+	maxUserTagLimit = 50
+
+	// tagRequestEndPoint is the endpoint for tag requests.
+	tagRequestEndPoint = "https://cloudresourcemanager.googleapis.com"
+
+	// projectParentPathFmt is the format string for parent path of a project resource.
+	projectParentPathFmt = "//cloudresourcemanager.googleapis.com/projects/%s"
+)
+
+// effectiveTags is the custom type for holding tags.
+type effectiveTags map[string]struct{}
+
+// processedUserTags is for storing the validated and processed tags
+// in-memory for later use. Validating and converting the tags to
+// required format involves using same GCP APIs and to reduce the
+// number of calls, making use of the in-memory storage approach.
+type processedUserTags struct {
+	processed bool
+	tags      map[string]string
+
+	sync.Mutex
+}
+
+// processedTags is the global instance for storing the processed tags.
+var processedTags = newProcessedUserTags()
+
+// tagManager handles resource tagging.
+type tagManager struct {
+	client API
+}
+
+// TagManager is the interface that wraps methods for resource tag operations.
+type TagManager interface {
+	getTagClientOptions() []option.ClientOption
+}
+
+// tagValuesClient handles operations on tag value resources.
+type tagValuesClient struct {
+	*rscmgr.TagValuesClient
+}
+
+// tagValuesClient handles operations on tag binding resources.
+type tagBindingsClient struct {
+	*rscmgr.TagBindingsClient
+}
+
+// newProcessedUserTags is for initializing an instance of processedUserTags.
+func newProcessedUserTags() *processedUserTags {
+	return &processedUserTags{}
+}
+
+// make is for initializing the internals of processedUserTags.
+func (p *processedUserTags) make(size int) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.tags = make(map[string]string, size)
+}
+
+// addTag is for adding a tag-set to processedUserTags.
+func (p *processedUserTags) addTag(k, v string) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.tags[k] = v
+}
+
+// updateState is for updating the tags processed status.
+func (p *processedUserTags) updateState() {
```

Comment:
```suggestion
func (p *processedUserTags) markProcessed() {
```
Or something like that.